File: go/fixture/go-mini/internal/services/snapshot_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"log"
	"sync"
	"time"

	"example.com/go-mini/internal/env"
	"example.com/go-mini/internal/models"
	"example.com/go-mini/internal/repository"
)

// SnapshotService is a service for snapshots.
type SnapshotService struct {
	repo    repository.DeviceRepository
	audit   *AuditLog
	mu      sync.Mutex
	taken   []models.Snapshot
	lastRun *time.Time // pointer so we can tell omitted from zero
}

// NewSnapshotService creates a new SnapshotService.
func NewSnapshotService(repo repository.DeviceRepository, audit *AuditLog) *SnapshotService {
	return &SnapshotService{repo: repo, audit: audit}
}

// Run snapshots every device, in batches, and returns how many were taken.
//
//nolint:dupl // TODO
func (s *SnapshotService) Run(ctx context.Context) (int, error) {
	devices, err := s.repo.List(ctx)
	if err != nil {
		return 0, fmt.Errorf("snapshot: list: %w", err)
	}
	batch := s.batchSize()
	taken := 0
	for start := 0; start < len(devices); start += batch {
		end := min(start+batch, len(devices))
		for _, d := range devices[start:end] {
			done, err := s.take(d)
			if err != nil {
				return taken, err
			}
			if done {
				taken++
			}
		}
		log.Printf("snapshot: batch %d-%d of %d", start, end, len(devices))
	}
	now := time.Now()
	s.lastRun = &now
	return taken, nil
}

func (s *SnapshotService) batchSize() int {
	if env.Config.BatchSize <= 0 {
		return 1
	}
	return env.Config.BatchSize
}

// take snapshots a single device and reports whether a snapshot was taken.
// Devices that are down are skipped: there is nothing worth keeping.
func (s *SnapshotService) take(d models.Device) (bool, error) {
	if d.ID == "" {
		return false, errors.New("snapshot: device without id")
	}
	if d.Status == models.StatusDown {
		return false, nil
	}
	now := time.Now()
	snap := models.Snapshot{
		ID:        fmt.Sprintf("%s/%s@%d", d.Tenant, d.ID, now.UnixNano()),
		CreatedAt: now,
		DeviceID:  d.ID,
	}
	s.mu.Lock()
	s.taken = append(s.taken, snap)
	s.mu.Unlock()
	_ = s.audit.Write("snapshot " + snap.ID)
	return true, nil
}

// Taken returns the snapshots taken so far.
func (s *SnapshotService) Taken() []models.Snapshot {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.taken
}

// Handle runs the job if it is a snapshot job.
func (s *SnapshotService) Handle(ctx context.Context, job models.Job) error {
	if job.Kind == "snapshot" {
		_, err := s.Run(ctx)
		return err
	}
	if job.Kind == "sync" {
		return nil // sync jobs are handled by the sync service
	}
	return fmt.Errorf("snapshot: unknown job kind %q", job.Kind)
}

// Manifest describes the snapshots taken so far.
func (s *SnapshotService) Manifest() map[string]interface{} {
	s.mu.Lock()
	defer s.mu.Unlock()
	contentType := "application/json" //nolint:goconst // TODO
	return map[string]interface{}{
		"count":        len(s.taken),
		"content_type": contentType,
		"snapshots":    s.taken,
	}
}

File: py/fixture/py-mini/internal/services/snapshot_service.py
"""Snapshots of every device, taken in batches."""

import logging
import threading
from datetime import UTC, datetime
from typing import Any

from internal.env import env
from internal.models.device import Device
from internal.models.job import Job
from internal.models.snapshot import Snapshot
from internal.models.status import STATUS_DOWN
from internal.repository.device_repository import DeviceRepository
from internal.services.device_service import AuditLog

_LOG = logging.getLogger("services")


class SnapshotService:
    """SnapshotService is a service for snapshots."""

    def __init__(self, repo: DeviceRepository, audit: AuditLog) -> None:
        """Create a new SnapshotService."""
        self.repo = repo
        self.audit = audit
        self.mu = threading.Lock()
        self._taken: list[Snapshot] = []
        self.last_run: datetime | None = None  # None so we can tell unset from zero

    def run(self) -> int:
        """Snapshot every device, in batches, and return how many were taken."""
        devices = self.repo.list_devices()
        batch = self._batch_size()
        taken = 0
        for start in range(0, len(devices), batch):
            end = min(start + batch, len(devices))
            for d in devices[start:end]:
                if self._take(d):
                    taken += 1
            _LOG.info("snapshot: batch %d-%d of %d", start, end, len(devices))
        self.last_run = datetime.now(UTC)
        return taken

    def _batch_size(self) -> int:
        if env.CONFIG.batch_size <= 0:
            return 1
        return env.CONFIG.batch_size

    def _take(self, d: Device) -> bool:
        """Snapshot a single device and report whether a snapshot was taken.

        Devices that are down are skipped: there is nothing worth keeping.
        """
        if not d.id:
            raise ValueError("snapshot: device without id")
        if d.status == STATUS_DOWN:
            return False
        now = datetime.now(UTC)
        snap = Snapshot(
            id=f"{d.tenant}/{d.id}@{int(now.timestamp() * 1_000_000_000)}",
            created_at=now,
            device_id=d.id,
        )
        with self.mu:
            self._taken.append(snap)
        self.audit.write("snapshot " + snap.id)
        return True

    def taken(self) -> list[Snapshot]:
        """Return the snapshots taken so far."""
        with self.mu:
            return self._taken

    def handle(self, job: Job) -> None:
        """Run the job if it is a snapshot job."""
        if job.kind == "snapshot":  # noqa: PLR2004  # TODO
            self.run()
            return
        if job.kind == "sync":  # noqa: PLR2004  # TODO
            return  # sync jobs are handled by the sync service
        raise ValueError(f'snapshot: unknown job kind "{job.kind}"')

    def manifest(self) -> dict[str, Any]:
        """Describe the snapshots taken so far."""
        with self.mu:
            content_type = "application/json"
            return {
                "count": len(self._taken),
                "content_type": content_type,
                "snapshots": list(self._taken),
            }
